### PR TITLE
refactor(game): extract SubAgentFSM from GameEngine (Issue #51 Phase 1b)

### PR DESCRIPTION
**Pull request by SweetSophia**
## SubAgentFSM Extraction — Pure Planner Pattern

This PR continues the GameEngine decomposition (Phase 1b of #51) by isolating the sub-agent lifecycle logic into a dedicated, side-effect-free module.

### What Changed

**New module: `src/game/SubAgentFSM.ts`**
A pure planner that computes the per-frame sub-agent state transition. It takes the character's current `spawnTime`, `dying`, and `fadeAlpha`, plus the current time and delta-seconds, and returns:
- `fadeAlpha` — new opacity value (clamped at 0)
- `dying` — new dying flag
- `actions` — list of one-shot side effects to perform (e.g., `spawn-despawn-sound`)
- `shouldRemove` — whether the caller should delete the character

The module also exports the `SUBAGENT_LIFETIME` (15s) and `SUBAGENT_FADE_DURATION` (2s) constants that were previously local to GameEngine.

**`GameEngine.update()` — refactored**
The previous inline lifecycle block (age check, despawn sound trigger, fade-out math, character deletion) is replaced with a call to `tickSubAgent`. The engine now applies the returned values to the character, executes any reported actions via `sfx.despawn()`, and removes the character when `shouldRemove` is true.

### Key Design Choice

The planner is **deliberately pure** — it returns action descriptors rather than mutating state or invoking `sfx` directly. This keeps the FSM unit-testable (no mocks needed) while keeping all side effects consolidated in GameEngine where they belong.

### Behavior Edge Cases Preserved

- **One-shot despawn sound**: The `justEnteredDying` guard ensures the sound fires exactly once on the age threshold crossing, and is *not* re-emitted when dying was set externally via `killSubAgent`.
- **Strict `>` boundary**: The despawn transition triggers only when `age > SUBAGENT_LIFETIME` (not `>=`), matching prior behavior.
- **Fade clamping**: `fadeAlpha` cannot go negative; removal happens exactly when fade reaches 0.
- **Kill-driven death**: When `char.dying` is already `true` (from `killSubAgent`), the sound is suppressed but fade-out proceeds normally.

### Test Coverage

`src/game/SubAgentFSM.test.ts` adds 10 unit tests covering:
- Young (non-dying) sub-agent defaults
- Age-threshold transition into dying with sound emission
- No re-emission of sound on subsequent ticks
- Fade-rate calculation matching `SUBAGENT_FADE_DURATION`
- Clamping at 0 with large `dt`
- Removal flag at zero alpha
- `killSubAgent`-style pre-dying input
- Exact boundary case at `LIFETIME` (no transition)
- Constant value verification

### API Impact

None. `spawnSubAgent` and `killSubAgent` remain public on GameEngine; PixelOffice callers see no change. All 175 tests pass and the build is green.

---

This pull request extracts the sub-agent finite state machine (FSM) logic from `GameEngine` into a dedicated `SubAgentFSM` module, completing **Phase 1b** of Issue #51's refactor plan.

## Key changes:

**1. `SubAgentFSM.ts`**
- Renames the `SubAgentAction` kind from `'spawn-despawn-sound'` → `'despawn-sound'` for clarity (this action only fires on despawn, not spawn).
- Preserves existing `fadeAlpha` for non-dying sub-agents instead of implicitly resetting it to `1`, matching the original `GameEngine` behavior. Previously, a non-dying branch was returning `fadeAlpha: 1`, which would have silently overwritten any partially-faded alpha. A regression test now pins this behavior.
- Introduces a shared `EMPTY_ACTIONS` constant to avoid allocating empty arrays on every tick.
- Adds an explanatory comment documenting the rationale for preserving `fadeAlpha`.

**2. `GameEngine.ts`**
- Updates the import to reference the renamed action kind (`'despawn-sound'`).
- Moves the `tickSubAgent` import to the top of the file alongside other module imports (it was previously imported mid-file).

**3. `SubAgentFSM.test.ts`**
- Updates the existing test to assert the renamed action kind (`'despawn-sound'`).
- Adds a new regression test verifying that `fadeAlpha` is preserved for non-dying sub-agents.
- Fixes the missing trailing newline at end of file.

## Functional impact:
- `SubAgentFSM` is now a self-contained, pure module — `tickSubAgent` takes an input state and returns a `SubAgentTick` result with no side effects beyond the explicit `actions` list.
- `GameEngine` no longer needs to know about sub-agent state-transition details; it simply applies the returned tick to the character.
- The behavior change to preserve `fadeAlpha` for non-dying sub-agents is a no-op against the original code (which also didn't reset it), now codified and protected by a test.

This is a pure refactor — no user-facing behavior changes, but it improves testability, isolates sub-agent lifecycle logic, and sets up further FSM extractions in subsequent phases of Issue #51.

### CODE REVIEW

File: src/game/SubAgentFSM.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  if (!dying) {
+    return { fadeAlpha: 1, dying: false, actions: [], shouldRemove: false };
+  }
+
+  const fadeAlpha = Math.max(0, char.fadeAlpha - dtSec / (SUBAGENT_FADE_DURATION / 1000));
+  const shouldRemove = fadeAlpha <= 0;
+  const actions: SubAgentAction[] = justEnteredDying ? [{ kind: 'spawn-despawn-sound' }] : [];
+
+  return { fadeAlpha, dying, actions, shouldRemove };
+}
```
[the remark this answers is not in the document]
Comment (SweetSophia) [reply src/game/SubAgentFSM.ts]: ✅ **F1 — VALID (medium) — FIXED in **

Introduced module-level  constant. Both the  early-return and the  branch now reuse it instead of allocating a fresh  every frame. This is a hot-path allocation (one call per sub-agent character per frame), so the fix matters.

File: src/game/GameEngine.ts [hunk 2]
```diff
@@ -76,8 +76,7 @@ const AGENT_PALETTES: Record<'cybera' | 'main' | 'shodan' | 'cyberlogis' | 'desc
   chi: 4, cylena: 5, sysauxilia: 3, miku: 0,
 };
 
-const SUBAGENT_LIFETIME = 15000; // ms before sub-agents fade out
-const SUBAGENT_FADE_DURATION = 2000; // fade animation length
+import { tickSubAgent } from './SubAgentFSM';
```
[the remark this answers is not in the document]
Comment (SweetSophia) [reply src/game/GameEngine.ts]: ✅ **F2 — VALID (medium) — FIXED in **

Moved  from line 79 to the top-of-file import block (now line 24, after the  type import). The duplicate mid-file statement was removed.  rule is now satisfied.

File: src/game/SubAgentFSM.test.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
+import { describe, it, expect } from 'vitest';
+import { tickSubAgent, SUBAGENT_LIFETIME, SUBAGENT_FADE_DURATION } from './SubAgentFSM';
+import type { SubAgentTick } from './SubAgentFSM';
```
[the remark this answers is not in the document]
Comment (SweetSophia) [reply src/game/SubAgentFSM.test.ts]: ✅ **F3 — VALID (low) — FIXED in **

Removed the unused  from the test file. The tests use the inferred return type of  and don't need to name  directly.

File: src/game/SubAgentFSM.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+export const SUBAGENT_LIFETIME = 15_000;
+export const SUBAGENT_FADE_DURATION = 2_000;
+
+export interface SubAgentInput {
+  spawnTime: number;
+  dying: boolean;
+  fadeAlpha: number;
+}
+
+export type SubAgentAction = { kind: 'spawn-despawn-sound' };
```
[the remark this answers is not in the document]
Comment (SweetSophia) [reply src/game/SubAgentFSM.ts]: ✅ **F4 — VALID (low) — FIXED in **

Renamed the action kind from  to . The action only ever fires on despawn transition (the spawn path calls  directly in , not through the FSM). The misleading prefix is gone. Updated the test assertion to match.

File: src/game/SubAgentFSM.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  char: SubAgentInput,
+  nowMs: number,
+  dtSec: number,
+): SubAgentTick {
+  const age = nowMs - char.spawnTime;
+  const dying = char.dying || age > SUBAGENT_LIFETIME;
+  const justEnteredDying = dying && !char.dying;
+
+  if (!dying) {
+    return { fadeAlpha: 1, dying: false, actions: [], shouldRemove: false };
```
[the remark this answers is not in the document]
Comment (SweetSophia) [reply src/game/SubAgentFSM.ts]: ✅ **F5 — VALID (low) — FIXED in **

Good catch. The original  block in  had no  branch setting , so non-dying sub-agents retained whatever value they had. My extraction was unconditionally returning  for the  case — a defensive change that could mask future bugs if someone adds a path that sets  to a non-1 value.

**Fix:** return  in the  branch, matching the original semantics exactly. Added a regression test () that pins this with a non-1 input value (0.7) to prevent future re-introduction of the implicit reset.

In practice this is a no-op (spawn sets , only the dying branch decreases it, and  only transitions ), but you're right that the implicit reset is the kind of change that bites later. Preserving original behavior is the correct call.

File: src/game/SubAgentFSM.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (SweetSophia) [reply src/game/SubAgentFSM.ts]: ✅ **F6 — VALID (low) — FIXED in **

Trailing newline added to . POSIX-compliant now.

File: src/game/SubAgentFSM.test.ts [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+});
+
+describe('SUBAGENT_LIFETIME and SUBAGENT_FADE_DURATION constants', () => {
+  it('SUBAGENT_LIFETIME is 15000ms', () => {
+    expect(SUBAGENT_LIFETIME).toBe(15_000);
+  });
+  it('SUBAGENT_FADE_DURATION is 2000ms', () => {
+    expect(SUBAGENT_FADE_DURATION).toBe(2_000);
+  });
+});
```
[the remark this answers is not in the document]
Comment (SweetSophia) [reply src/game/SubAgentFSM.test.ts]: ✅ **F7 — VALID (low) — FIXED in **

Trailing newline added to . Both files in this PR now have trailing newlines.

File: src/game/SubAgentFSM.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (SweetSophia) [reply src/game/SubAgentFSM.ts]: ✅ **F1 — VALID (medium) — FIXED in `fc931ea`**

Introduced module-level `EMPTY_ACTIONS: SubAgentAction[] = []` constant. Both the `!dying` early-return and the `already-dying` branch now reuse it instead of allocating a fresh `[]` every frame. This is a hot-path allocation (one call per sub-agent character per frame), so the fix matters.

```ts
const EMPTY_ACTIONS: SubAgentAction[] = [];
// ...
if (!dying) {
  return { fadeAlpha: char.fadeAlpha, dying: false, actions: EMPTY_ACTIONS, shouldRemove: false };
}
// ...
const actions: SubAgentAction[] = justEnteredDying
  ? [{ kind: 'despawn-sound' }]
  : EMPTY_ACTIONS;
```

✅ Resolving this thread.

File: src/game/GameEngine.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (SweetSophia) [reply src/game/GameEngine.ts]: ✅ **F2 — VALID (medium) — FIXED in `fc931ea`**

Moved `import { tickSubAgent } from './SubAgentFSM';` from line 79 to the top-of-file import block (now line 24, after the `PlacedFurniture` type import). The duplicate mid-file statement was removed. `import/first` rule is now satisfied.

✅ Resolving this thread.

File: src/game/SubAgentFSM.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (SweetSophia) [reply src/game/SubAgentFSM.ts]: ✅ **F4 — VALID (low) — FIXED in `fc931ea`**

Renamed the action kind from `'spawn-despawn-sound'` to `'despawn-sound'`. The action only ever fires on despawn transition (the spawn path calls `sfx.spawn()` directly in `spawnSubAgent`, not through the FSM). The misleading prefix is gone. Updated the test assertion to match.

✅ Resolving this thread.

File: src/game/SubAgentFSM.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (SweetSophia) [reply src/game/SubAgentFSM.ts]: ✅ **F5 — VALID (low) — FIXED in `fc931ea`**

Good catch. The original `if (char.isSubAgent) { ... }` block in `GameEngine.update()` had no `else` branch setting `fadeAlpha`, so non-dying sub-agents retained whatever value they had. My extraction was unconditionally returning `fadeAlpha: 1` for the `!dying` case — a defensive change that could mask future bugs if someone adds a path that sets `fadeAlpha` to a non-1 value.

**Fix:** return `fadeAlpha: char.fadeAlpha` in the `!dying` branch, matching the original semantics exactly. Added a regression test (`'preserves existing fadeAlpha for non-dying sub-agent (no implicit reset)'`) that pins this with a non-1 input value (0.7) to prevent future re-introduction of the implicit reset.

In practice this is a no-op (spawn sets `fadeAlpha: 1`, only the dying branch decreases it, and `dying` only transitions `false → true`), but you're right that the implicit reset is the kind of change that bites later. Preserving original behavior is the correct call.

✅ Resolving this thread.

File: src/game/SubAgentFSM.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (SweetSophia) [reply src/game/SubAgentFSM.ts]: ✅ **F6 — VALID (low) — FIXED in `fc931ea`**

Trailing newline added to `src/game/SubAgentFSM.ts`. POSIX-compliant now.

✅ Resolving this thread.

File: src/game/SubAgentFSM.test.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (SweetSophia) [reply src/game/SubAgentFSM.test.ts]: ✅ **F7 — VALID (low) — FIXED in `fc931ea`**

Trailing newline added to `src/game/SubAgentFSM.test.ts`. Both files in this PR now have trailing newlines.

✅ Resolving this thread.

**Comment by SweetSophia** [comment]
## Consolidated review — PR #76 (Phase 1b: SubAgentFSM extraction)

### Triage of all review activity

**Reviewers:** gemini-code-assist, copilot-pull-request-reviewer, kilo-code-review, kody-code-review, socket-security. Plus my independent self-review (oracle dispatch was corrupted by cross-session context contamination; self-review via codegraph + AFT + test pinning).

**7 inline threads + 3 bot summaries:**
| # | Author | Finding | Verdict |
|---|---|---|---|
| F1 | gemini | GC pressure from `[]` allocation per frame | **VALID (medium)** — fixed in `fc931ea` |
| F2 | gemini | Mid-file import violates `import/first` | **VALID (medium)** — fixed |
| F3 | copilot | Unused `SubAgentTick` import in test | **VALID (low)** — fixed (thread auto-resolved) |
| F4 | kilo | Action kind `spawn-despawn-sound` misnamed | **VALID (low)** — renamed to `despawn-sound` |
| F5 | kilo | `fadeAlpha:1` for non-dying sub-agent is a behavior change | **VALID (low)** — preserved original behavior, added regression test |
| F6 | kilo | `SubAgentFSM.ts` missing trailing newline | **VALID (low)** — fixed |
| F7 | kilo | `SubAgentFSM.test.ts` missing trailing newline | **VALID (low)** — fixed |

**Bot summaries:** Socket ✅ pass, Kody ✅ pass, Kilo ✅ pass. Sourcery skipping (platform-wide pattern, not code-related).

### Verification

```
typecheck: E0 W0
tests:     176/176 pass (11 SubAgentFSM tests, 1 new regression for F5)
build:     ✅
coverage:  GameEngine unchanged; SubAgentFSM 100% line coverage
```

### Semantic-anchor assessment

| Anchor | Before | After |
|---|---|---|
| **Pure-function extraction** | `tickSubAgent` logic interleaved with movement/animation in 15-line inline block | Isolated pure function with explicit `SubAgentTick` return shape |
| **Side-effect decoupling** | `sfx.despawn()` called inline; one-shot guard via `!char.dying` check | `tickSubAgent` emits `actions: [{ kind: 'despawn-sound' }]`; engine applies effects in same frame, preserving one-shot semantics |
| **State machine clarity** | `dying` flag + `fadeAlpha` + age check all entangled | `dying` derived from `char.dying OR age > LIFETIME`; `fadeAlpha` reduced by `dt / (FADE_DURATION / 1000)`; `shouldRemove` derived from `fadeAlpha <= 0` |
| **Hot-path allocation** | `[]` allocated per sub-agent per frame | `EMPTY_ACTIONS` module-level constant reused |

### Risk analysis

This PR carries **zero behavioral risk**:
1. **One-shot despawn sound:** preserved by `justEnteredDying = dying && !char.dying` check inside the planner
2. **Boundary strict-greater-than (`> 15000`):** preserved verbatim in `age > SUBAGENT_LIFETIME`
3. **Fade rate:** preserved (`dt / (SUBAGENT_FADE_DURATION / 1000)`)
4. **Non-dying sub-agent fadeAlpha:** now returns `char.fadeAlpha` (was implicitly preserved by absence of write; F5 fix makes the contract explicit + testable)
5. **Non-sub-agent fadeAlpha=1:** still set by the `else` branch in GameEngine (unchanged)

### What's deferred to Phase 2 (separate PR)

- **`EditorController` extraction** (mouse/touch/keyboard handlers) — Phase 2 of the GameEngine decomposition plan
- **Other `GameEngine.ts` 2086-line surface** — pathfinding, obstacle rebuild, monitor glow, speech bubbles, etc. Each will get its own targeted extraction PR with TDD tests

### Verdict

**Ready to merge.** All 7 review findings addressed in `fc931ea`. No outstanding threads. CI fully green. Behavior-preserving extraction with explicit test coverage of every state transition.

Closes issue #51 (Phase 1b; Phase 2 EditorController will be a follow-up).
